### DISCUSSION TITLE
can't connect the library to my project via Cmake

### DISCUSSION THREAD

**Discussion by safocl** [nlohmann/json#2230] [Q&A] [answered]
#### What is the issue you have?

can't connect the library to my project via Cmake as described here: https://github.com/nlohmann/json#embedded.
gives out an error https://i.imgur.com/ihLeJQn.png
`--> LC_ALL=C make -j4
[  7%] Built target configure
[ 14%] Building CXX object src/loger/CMakeFiles/loger.dir/loger.cpp.o
[ 21%] Built target timer
Scanning dependencies of target mainWindow
[ 28%] Built target aclock
In file included from /home/safff/Projects/Watcher/src/loger/loger.cpp:2:
/home/safff/Projects/Watcher/src/configure/configure.hpp:6:10: fatal error: nlohmann/json.hpp: No such file or directory
    6 | #include <nlohmann/json.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/loger/CMakeFiles/loger.dir/build.make:83: src/loger/CMakeFiles/loger.dir/loger.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1080: src/loger/CMakeFiles/loger.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 35%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/clockentity.cpp.o
[ 42%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/logentity.cpp.o
[ 50%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/mainform.cpp.o
[ 57%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/timerentity.cpp.o
[ 64%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/wfraclock.cpp.o
[ 71%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/wfrlog.cpp.o
[ 78%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/wfrtm.cpp.o
[ 85%] Building CXX object src/ui/CMakeFiles/mainWindow.dir/wnotebook.cpp.o
[ 85%] Built target mainWindow
make: *** [Makefile:150: all] Error 2
`

my project: https://github.com/safocl/Watcher/tree/tmp

#### Please describe the steps to reproduce the issue.

1.
2.
3.

#### Which compiler and operating system are you using?

- Compiler: gcc 10.1.0
- Operating system: archlinux

#### Which version of the library did you use?

- [ ] latest release version 3.7.3
- [ ] other release - please state the version: ___
- [*] the `develop` branch

#### If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

- [ ] yes
- [*] no - please copy/paste the error message below

**Comment by nlohmann** [comment]
(I moved the issue to the discussion board, because it's a question rather than a bug).

Browsing your code, it seems you are only adding the project, but not linking it to your binary, see https://github.com/safocl/Watcher/blob/3a500c8288cc9a8e36c13db9ea5ae3e856bb52a1/src/CMakeLists.txt#L19

**Reply by safocl** [reply]
this line was previously decomposed and had no effect.

**Reply by nlohmann** [reply]
Here is an example that is part of the tests: https://github.com/nlohmann/json/tree/develop/test/cmake_add_subdirectory/project

**Reply by safocl** [reply]
the same thing happens to this line:
https://i.imgur.com/Qo6eTLi.png

**Reply by nlohmann** [reply]
Please do not post screenshots, but rather copy/paste the error log.

**Comment by safocl** [comment]
```
`--> LC_ALL=C make -j4
[  7%] Building CXX object src/loger/CMakeFiles/loger.dir/loger.cpp.o
[ 14%] Built target configure
[ 21%] Built target timer
[ 28%] Built target aclock
[ 85%] Built target mainWindow
In file included from /home/safff/Projects/Watcher/src/loger/loger.cpp:2:
/home/safff/Projects/Watcher/src/configure/configure.hpp:6:10: fatal error: nlohmann/json.hpp: No such file or directory
    6 | #include <nlohmann/json.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/loger/CMakeFiles/loger.dir/build.make:83: src/loger/CMakeFiles/loger.dir/loger.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1080: src/loger/CMakeFiles/loger.dir/all] Error 2
make: *** [Makefile:150: all] Error 2

```

same mistake

**Reply by safocl** [reply]
and I'd like to point out that for https://github.com/safocl/Watcher/blob/3a500c8288cc9a8e36c13db9ea5ae3e856bb52a1/src/configure/configure.cpp, this line helped:
https://github.com/safocl/Watcher/blob/3a500c8288cc9a8e36c13db9ea5ae3e856bb52a1/src/configure/CMakeLists.txt#L9.

which is similar to the situation, but with a successful outcome.

if you remove this line, it works:
```
`--> LC_ALL=C make -j4
Scanning dependencies of target configure
[  7%] Building CXX object src/configure/CMakeFiles/configure.dir/configure.cpp.o
[ 14%] Building CXX object src/loger/CMakeFiles/loger.dir/loger.cpp.o
[ 21%] Built target aclock
[ 28%] Built target timer
In file included from /home/safff/Projects/Watcher/src/configure/configure.cpp:1:
/home/safff/Projects/Watcher/src/configure/configure.hpp:6:10: fatal error: nlohmann/json.hpp: No such file or directory
    6 | #include <nlohmann/json.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/configure/CMakeFiles/configure.dir/build.make:83: src/configure/CMakeFiles/configure.dir/configure.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1161: src/configure/CMakeFiles/configure.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 85%] Built target mainWindow
In file included from /home/safff/Projects/Watcher/src/loger/loger.cpp:2:
/home/safff/Projects/Watcher/src/configure/configure.hpp:6:10: fatal error: nlohmann/json.hpp: No such file or directory
    6 | #include <nlohmann/json.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/loger/CMakeFiles/loger.dir/build.make:83: src/loger/CMakeFiles/loger.dir/loger.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1080: src/loger/CMakeFiles/loger.dir/all] Error 2
make: *** [Makefile:150: all] Error 2

```

**Comment by safocl** [comment, marked as answer]
ох, это была моя ошибка -- я вместо target_link_libraries указал target_link_directories
https://github.com/safocl/Watcher/blob/3a500c8288cc9a8e36c13db9ea5ae3e856bb52a1/src/loger/CMakeLists.txt#L8

**Reply by nlohmann** [reply]
Why don't you use `target_link_libraries`?

**Reply by safocl** [reply]
> Why don't you use `target_link_libraries`?

yeah, I made up for my mistake and it all worked out.

**Reply by safocl** [reply]
thanks, this discussion can be closed.
